contracts/cryptoblades.sol

```diff
@@ -451,6 +460,60 @@ contract CryptoBlades is Initializable, AccessControlUpgradeable {
         return (((attacker + 1) % 4) == defender); // Thanks to Tourist
     }
 
+    function _updatePaymentBlockHash() internal {
+        if ((mintPayments[msg.sender].count > 0) &&
+            (mintPayments[msg.sender].blockHash == 0) &&
+            (mintPayments[msg.sender].blockNumber < block.number)) {
+
+            // Payment must be recent enough that the hash is available for the payment block.
+            // Use 200 as a 'friendly' window of "You have 10 minutes."
+            if (mintPayments[msg.sender].blockNumber + 200 >= block.number) {
+                mintPayments[msg.sender].blockHash = blockhash(mintPayments[msg.sender].blockNumber);
+            }
+        }
+    }
+
+    function _discardPaymentIfExpired() internal {
+        _updatePaymentBlockHash();
```

Comment:
we probably want to do the blockhash update in the catch part of the minting code?

Reply:
> we probably want to do the blockhash update in the catch part of the minting code?

I was thinking we want to update the blockhash as soon as possible.

```
    function mintWeapon() public onlyNonContract oncePerBlock(msg.sender)  {
        _discardPaymentIfExpired();
        try weapons.mint(msg.sender, randoms.getRandomSeed(msg.sender)) {
            _usePayment(address(weapons), 1);
        }
        catch {
        }
    }
```

My understanding is that the effects of _discardPaymentIfExpired() will not be reverted if the try fails.

If that's not the case, then we need to add redundant-looking logic to the catch:

```
    function mintWeapon() public onlyNonContract oncePerBlock(msg.sender)  {
        _discardPaymentIfExpired();
        try weapons.mint(msg.sender, randoms.getRandomSeed(msg.sender)) {
            _usePayment(address(weapons), 1);
        }
        catch {
            _discardPaymentIfExpired();
        }
    }
```